Updated TypeReference and TypeDecoder to support decoding of dynamic structs and dynamic struct arrays without a priori Class definitions.

Antlion12 opened the pull request:
- **Updated TypeReference object to support nested types to support truly dynamic stucts.**
- **Support decoding of DynamicArray of DynamicStruct (with corresponding unit test). Change visibility of TypeReference's innerTypes and getSubTypeReference() to protected/public to allow for overriding in anonymous classes. Redo naming of some innerType-specific functions in TypeDecoder to minimize code diffs.**

### What does this PR do?
This PR builds off of the PR originally created by calmacfadden at https://github.com/hyperledger/web3j/pull/2023.

### Where should the reviewer start?
* Look in TypeDecoder.java, with attention to the decodeDynamicParameterFromStructWithTypeReference() and
decodeDynamicStructElementsFromInnerTypes(). Note that the Ctor() version of functions from PR 2023 have been
renamed to omit the Ctor so as to match the name in the main branch and reduce code diffs.

* The unit test testArrayOfDynamicStruct() provides an example of decoding an array of dynamic structs (uint256,bool,string)[]
based on a transaction from the Treasure Ruby testnet.

* I've taken some stylistic liberty by renaming TypeReference.innerTypeReferences() to TypeReferences.innerTypes().

* I've changed the scope of some fields in TypeReference from private to protected/public to allow for creating the
anonymous DynamicArray class seen in testArrayOfDynamicStruct().

### Why is it needed?
As with PR 2023, it would be useful to decode dynamic structs without having to explicitly define a corresponding class.

## Checklist

- [x] I've read the contribution guidelines.
- [x] I've added tests (if applicable).
- [x] I've added a changelog entry if necessary.

gtebrean commented on abi/src/test/java/org/web3j/abi/DefaultFunctionEncoderTest.java:
```diff
@@ -70,6 +59,144 @@ public void testBuildMethodSignatureWithStructWithArray() {
+        // Transaction:
+        // https://testnet.treasurescan.io/tx/0x041e53e7571283d462df99a95b2c21324279657f26a3adef907095d2d9c5ed85?tab=logs
+        // Contract:
+        // https://testnet.treasurescan.io/address/0x5167E9A422aCEd95C2D0b62bF05a7847a9a942B2
```
please create a simpler contract example  an deploy it on sepolia as above example, this one is tricky because it shows first the interface and developers will need to search a bit more to find the struct

gtebrean replied:
I changed the test method name, it can be keep as it is

gtebrean commented on abi/src/main/java/org/web3j/abi/TypeDecoder.java:
```diff
@@ -464,7 +475,92 @@ public static <T extends Type> T decodeDynamicStruct(
+        final List<Integer> parameterOffsets = new ArrayList<>();
+
+        int dynamicParametersToProcess = 0;
+        for (int i = 0; i < length; ++i) {
```
extract this for in a separated method to be easier to follow as this is computing the offsets

Antlion12 replied:
Change has been made in  https://github.com/hyperledger/web3j/pull/2076/commits/a08ac8470d59272a4f6bf9e43daa9fe398555720.

gtebrean commented on abi/src/main/java/org/web3j/abi/TypeDecoder.java:
```diff
@@ -464,7 +475,92 @@ public static <T extends Type> T decodeDynamicStruct(
+            }
+        }
+        int dynamicParametersProcessed = 0;
+        for (int i = 0; i < length; ++i) {
```
extract this for in another method, as this is adding the dynamic values to parameter list, it will be easier to watch

Antlion12 replied:
Change also made in https://github.com/hyperledger/web3j/pull/2076/commits/a08ac8470d59272a4f6bf9e43daa9fe398555720.

gtebrean reviewed (changes requested):
please add 3 more test cases except these to ensure the functionality and split some functionalities in different methods for an easier tracking of the flow

Antlion12 commented on abi/src/main/java/org/web3j/abi/TypeDecoder.java:
```diff
@@ -464,7 +475,92 @@ public static <T extends Type> T decodeDynamicStruct(
+                && typeReference.getInnerTypes() != null) {
+            return decodeDynamicStructElementsFromInnerTypes(
+                    input, offset, typeReference, function);
+        } else {
```
Thanks. Fixed in this commit: https://github.com/hyperledger/web3j/pull/2076/commits/290bba705dc9dd8cece455dd23462fb48a34a358

Antlion12 commented:
> please add 3 more test cases except these to ensure the functionality and split some functionalities in different methods for an easier tracking of the flow

Sounds good. I'll chip away at these.

Question about the "add 3 more test cases" request: the test `testDynamicStructFix()` covers dynamic struct inside a dynamic struct. The `testArrayOfDynamicStruct()` covers an array of dynamic struct. What other permutations are missing?

gtebrean commented:
> > please add 3 more test cases except these to ensure the functionality and split some functionalities in different methods for an easier tracking of the flow
> 
> Sounds good. I'll chip away at these.
> 
> Question about the "add 3 more test cases" request: the test `testDynamicStructFix()` covers dynamic struct inside a dynamic struct. The `testArrayOfDynamicStruct()` covers an array of dynamic struct. What other permutations are missing?

Sorry I forgot to mention, add dynamic struct of static struct, dynamic struct with type reference of dynamic arrays and static struct of dynamic struct in order to make sure that different use cases were covered

gtebrean commented:
@Antlion12 I tried to add the required changes on this PR in order to can include it in the next release but on the first required test case I notice that the DynamicStruct of StaticStruct is not supported, you can notice in the added test that the decode is failing. 
Meantime you find the time to get back to it I will also try to find a fix when I have the time.

Antlion12 commented:
To fix the unit test, I have added support for building StaticStruct with innerTypes. I have also added unit tests of DynamicStruct of StaticStruct (built with inner types), as well as DynamicArray of StaticStruct (built with inner types).

gtebrean commented:
I agree with the changes, from what I see only the unit test StaticStruct of DynamicStruct is missing. Once you commit that I will merge this PR.

Antlion12 commented:
> I agree with the changes, from what I see only the unit test StaticStruct of DynamicStruct is missing. Once you commit that I will merge this PR.

Does a StaticStruct of DynamicStruct need to be tested if it's not considered valid ABI? I was looking at this part of the [ABI encoding spec](https://docs.soliditylang.org/en/latest/abi-spec.html#formal-specification-of-the-encoding) and saw 


> Definition: The following types are called “dynamic”:
> ...
> * `T[k]` for any dynamic `T` and any `k >= 0`
> * `(T1,...,Tk)` if `Ti` is dynamic for some `1 <= i <= k`

So if there was an outer struct that contained a DynamicStruct inside of it, then that outer struct can only be Dynamic (it couldn't be Static).

Likewise, a StaticArray of a DynamicStruct is not valid ABI, since `T` being dynamic impleis `T[k]` is dynamic.

That's my understanding of it, at least.

gtebrean commented:
> ABI encoding spec

Indeed, I omitted this, my bad.

gtebrean commented on abi/src/main/java/org/web3j/abi/TypeDecoder.java:
```diff
@@ -371,9 +377,68 @@ public static <T extends Type> T decodeStaticStruct(
     }
 
+    // Counts the number of nested fields in a StaticStruct with inner types.
+    private static <T extends Type> int numNestedFields(final TypeReference<T> typeReference) {
```
```suggestion
    private static <T extends Type> int countNestedFields(final TypeReference<T> typeReference) {
```

Antlion12 commented:
I have renamed `numNestedFields()` to `countNestedFields()`, and i have updated the CHANGELOG.md to refer to this PR using the description "Add struct support in java without the need of having a corresponding Java class " provided from the 2024--2025 roadmap (https://wiki.hyperledger.org/display/WEB3J/Roadmap+2024+-+2025).
